synapse_token_authenticator/token_authenticator.py

```diff
@@ -34,26 +42,56 @@ class TokenAuthenticator(object):
     def __init__(self, config: dict, account_handler: ModuleApi):
         self.api = account_handler
 
-        self.config = config
-        if self.config.secret:
-            k = {
-                "k": base64.urlsafe_b64encode(
-                    self.config.secret.encode("utf-8")
-                ).decode("utf-8"),
-                "kty": "oct",
-            }
-            self.key = jwk.JWK(**k)
-        else:
-            with open(self.config.keyfile, "r") as f:
-                self.key = jwk.JWK.from_pem(f.read())
-
-        self.api.register_password_auth_provider_callbacks(
-            auth_checkers={
-                ("com.famedly.login.token", ("token",)): self.check_auth,
-            },
-        )
+        auth_checkers = {}
+
+        self.config: TokenAuthenticatorConfig = config
+        if (jwt := getattr(self.config, "jwt", None)) is not None:
+            if jwt.secret:
+                k = {
+                    "k": base64.urlsafe_b64encode(jwt.secret.encode("utf-8")).decode(
+                        "utf-8"
+                    ),
+                    "kty": "oct",
+                }
+                self.key = jwk.JWK(**k)
+            else:
+                with open(jwt.keyfile, "r") as f:
+                    self.key = jwk.JWK.from_pem(f.read())
+            auth_checkers[("com.famedly.login.token", ("token",))] = self.check_jwt_auth
+
+        if (oidc := getattr(self.config, "oidc", None)) is not None:
+            auth_checkers[
+                ("com.famedly.login.token.oidc", ("token",))
+            ] = self.check_oidc_auth
+
+            self.api.register_web_resource(
+                "/_famedly/login/com.famedly.login.token.oidc",
+                self.LoginMetadataResource(oidc),
+            )
+
+        self.api.register_password_auth_provider_callbacks(auth_checkers=auth_checkers)
 
-    async def check_auth(
+    class LoginMetadataResource(resource.Resource):
+        def __init__(self, oidc_config: object):
+            self.issuer = oidc_config.issuer
+            self.metadata_url = urljoin(
+                oidc_config.issuer, "/.well-known/openid-configuration"
+            )
+            self.organization_id = oidc_config.organization_id
+            self.project_id = oidc_config.project_id
+
+        def render_GET(self, request):
+            request.responseHeaders.addRawHeader(b"content-type", b"application/json")
```

Comment:
This looks like some changes were reverted?

```suggestion
            request.setHeader(b"content-type", b"application/json")
            request.setHeader(b"access-control-allow-origin", b"*")
```

Reply:
Fixed